README
NAME
    Device::MX240 - Perl api to the MX240 (Motorola Instant Messenger)

SYNOPSIS
            use Device::MX240 qw(open_dev);

            my $done = 0;

            # Ctrl-C/INT handler
            $SIG{INT} = sub { $SIG{INT} = 'IGNORE'; $done++; };

            my $d = open_dev();

            # init device
            $d->write_dev("\xad\xef\x8d");

            while (!$done) {
                    my $buf = $d->read_dev();
                    if ($buf) {
                            # process $buf
                            # and use $d->write_dev();
                    }
                    sleep(1);
                    # poll the device
                    $d->write_dev("\xad");
            }

            $d->close_dev;

            exit;

ABSTRACT
    This module allows easy communication to and from the MX240. Also known
    as the IMFree. The device is a wireless instant messenger. (USB base,
    900mhz) libusb and libhid are required for this module to compile.

DESCRIPTION
    This module has a few functions to interact with the MX240a. Data is
    read in 16 byte boundries, and is written in 16 byte boundries.

  CONSTRUCTOR
    "open_dev"
        This function returns a blessed reference to the MX240 device. The
        interface for this function may change in the future to accept a
        vendor and product id. You can use the returned object with the
        following mehods.

  METHODS
    "read_dev"
                Returns undef when nothing has been read, and 16 bytes of data on success.

    "write_dev( $data )"
                Returns number of bytes written, and 0 on failure.

    "close_dev"
                Returns true on success and 0 on failure;

  EXPORTS
    None by default. You can export "open_dev", "MX240A_VENDOR", and
    "MX240A_PRODUCT".

            use Device::MX240 qw(open_dev);

SEE ALSO
    <http://libusb.sourceforge.net/>, <http://libhid.alioth.debian.org/>

AUTHOR
    David Davis <xantus@cpan.org>

CREDITS
    Dusty, for the C functions to libhid and libusb.

    Please rate this module.
    "http://cpanratings.perl.org/rate/?distribution=Device-MX240"

COPYRIGHT AND LICENSE
    Copyright 2005 by David Davis and Teknikill Software

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Device::MX240

0.04  Tue May 10 01:34:00 2005
	- working version!
	- simple example added
	- some docs added

0.01  Wed May 04 21:10:00 2005
	- original version


Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
	'NAME'		=> 'Device::MX240',
	'VERSION_FROM'	=> 'MX240.pm',
	'PREREQ_PM'		=> {
		# module => '1.01',
	},
	'dist'         => {
		'SUFFIX'       => ".gz",
		'DIST_DEFAULT' => 'all tardist',
		'COMPRESS'     => "gzip -9vf",
	},
	($] >= 5.005 ?    ## Add these new keywords supported since 5.005
		(
			ABSTRACT_FROM => 'MX240.pm', # retrieve abstract from module
			AUTHOR     => 'David Davis <xantus@cpan.org>') : ()),
			'LIBS'		=> ['-lhid -lusb'],
			'DEFINE'		=> '', # e.g., '-DHAVE_SOMETHING'
			# Insert -I. if you add *.h files later:
#			'INC'		=> '-I/usr/include/', # e.g., '-I/usr/include/other'
			# Un-comment this if you add C files to link with later:
#			'OBJECT'		=> '$(O_FILES)', # link all the C files too
);


examples/simple.pl
#!/usr/bin/perl

# lame test script!

use strict;

my $unit = 1;

# spaces are auto removed
my $INIT = "ad ef 8d 00 00 00 00 00 00 00 00 00 00 00 00 00";
my $POLL = "ad 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00";
my $SRVN = "c%d d7 20 41 49 4d 20 20 ff 00 00 00 00 00 00 00";

use Device::MX240;
use bytes;

my $done = 0;

$SIG{INT} = sub {
	print "closing\n";
	$SIG{INT} = 'IGNORE';
	$done = 1;
};

my $d = Device::MX240::open_dev();

sleep(2);
print "device init...";
my $i = $d->write_dev(hex2bin($INIT));
print STDERR "out: $INIT\n";
sleep(1);

my $hid;
my $msg = '';
my $m = 0;
while (!$done) {
	my $buf = $d->read_dev();
	my $len = length($buf);
	my $r;
	if ($len > 0) {
		print "len:$len\n";
		my $h = bin2hex($buf);
		# stupid case
		$h = "f$m $h" if ($m);
		print STDERR "in: ".bin2hex($buf)."\n";
		if ($h =~ m/^e\d fd/) {
			print "ack\n";
		} elsif ($h =~ m/^f\d fd/) {
			print "unknown\n";
		} elsif ($h =~ m/^f\d 8e ff/) {
			# disconnect
			print "disconnect\n";
		} elsif ($h =~ m/^f\d 8e/) {
			# connect
			print "connect\n";
			$r = $d->write_dev(hex2bin(sprintf($SRVN,$unit)));
			print STDERR "out: ".sprintf($SRVN,$unit)." <- srvname\n";
#			$unit++;
		} elsif ($h =~ m/^f\d 91/) {
			# user
			print "user\n";
		} elsif ($h =~ m/^f(\d) 92/) {
			# pass
			print "pass\n";
			# login ok
			$r = $d->write_dev(hex2bin("e$1 d3 00 00 00 00 00 00 00 00 00 00 00 00 00 00"));
			print STDERR "out: e0 d3 00 00 00 00 00 00 00 00 00 00 00 00 00 00 <- login ok\n";
			sleep(1);
			# newpsn
			# this may be wrong..
			# e1 ca 41 4e 4e 01 00 00
			$r = $d->write_dev(hex2bin("e$1 ca 41 4e 4e 01 00 00 00 00 00 00 00 00 00 00"));
			print STDERR "out: e0 ca 41 4e 4e 01 00 00 00 00 00 00 00 00 00 00 <- newpsn\n";
			# psndata TEST
			# MPD data:
			# c1 c9 4d 50 44 20 20 20
			# 4d 50 44 ff 00 00 00 00
			$r = $d->write_dev(hex2bin("c$1 c9 54 45 53 54 20 20 54 45 53 54 20 00 00 00"));
			print STDERR "out: c$1 c9 54 45 53 54 20 20 54 45 53 54 20 00 00 00 <- psndata\n";	
			
		} elsif ($h =~ m/^f(\d) 94/) {
			# talk
			print "talk!! $1\n";
			$hid = $1;
			my $dat = hex2bin("8$hid 01 00 ".bin2hex("hello")." ff");
			print "out: $dat : ".bin2hex($dat)."\n";
			$d->write_dev($dat);
		} elsif ($h =~ m/^f\d 95/) {
			# bye
			print "bye\n";
		} elsif ($h =~ m/^0e ce/) {
			# logoff
			print "logoff\n";
		} elsif ($h =~ m/^f\d 9b/) {
			# chat
			print "chat\n";
		} elsif ($h =~ m/^e0 ce/) {
			# pres
			print "pres\n";
			$r = $d->write_dev(hex2bin("ee d3 00 00 00 00 00 00 00 00 00 00 00 00 00 00"));
			print STDERR "out: ee d3 00 00 00 00 00 00 00 00 00 00 00 00 00 00 <- ? pres ok\n";
			sleep(1);
			$r = $d->write_dev(hex2bin(sprintf($SRVN,$unit)));
			print stderr "out: ".sprintf($SRVN,$unit)." <- srvname\n";
			$unit++;
		} elsif ($h =~ m/^f\d 8c/) {
			# connect fail
			print "connect fail\n";
		} elsif ($hid && $h =~ m/^f(\d)(.*)/) {
			my $hid = $1;
			# d1 31 31 31 31 31 31 31 31 31 31 31 31 31 31 31
			# 31 31 31 31 31 31 31 31 31 31 31 31 31 31 31 fe
			# d1 31 31 31 31 31 31 31 31 31 31 31 31 31 31 31
			# 31 31 31 31 31 31 31 31 31 31 31 31 31 31 31 fe
			# f1 31 31 31 31 31 31 31 31 31 31 ff fe 00 31 31
			# msgin / not ff?
			print "msg in\n";
			my $txt = $2;
			if ($txt =~ s/ff.*//) {
				$m = 0;
				$txt = $msg.hex2bin($txt);
			} else {
				$m = $hid;
				$msg .= hex2bin($txt);
				next;
			}
			
			print "msg: [".$txt."]\n";
			# msgout:N|[id]:[warn]:[message]
			# 81 02 00 68 65 6c 6c 6f
			# 20 77 6f 72 6c 64 ff
			my $dat = hex2bin("8$hid 01 00 ".bin2hex($txt)." ff 00");
			print "out: $dat : ".bin2hex($dat)."\n";
			$d->write_dev($dat);
		} elsif ($h =~ m/^ef 01 01 fe/) {
			# init ok
			print "init ok\n";
		}
		
		if ($r) { print "write: $r\n"; }
	}
	sleep(1);
	my $n = $d->write_dev(hex2bin($POLL));
#	print STDERR "out: $POLL\n";
#	print "poll write: $n\n";
}

$d->close_dev();

print "clean exit\n";

exit 0;

sub hex2bin {
	my $todecode = shift;
	$todecode =~ s/\s+//g;
	$todecode =~ s/([0-9a-fA-F]{2})/pack("c",hex($1))/ge;
	return $todecode;
}

sub bin2hex {
	my ($d) = @_;
	$d =~ s/(.)/sprintf("%02x ",ord($1))/egs;
	return $d;
}



MX240.pm
package Device::MX240;

#use 5.006;
use strict;
use warnings;
use Errno;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

our $VERSION = '0.04';

our @ISA = qw(Exporter DynaLoader);

# This allows declaration use Device::MX240 ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	open_dev
	MX240A_VENDOR
	MX240A_PRODUCT
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw( );

sub AUTOLOAD {
	# This AUTOLOAD is used to 'autoload' constants from the constant()
	# XS function.  If a constant is not found then control is passed
	# to the AUTOLOAD in AutoLoader.

	my $constname;
	our $AUTOLOAD;
	($constname = $AUTOLOAD) =~ s/Device::MX240:://;
	croak "&$_[0] not defined in Device::MX240" if $constname eq 'constant';
	my $val = constant($constname, @_ ? $_[0] : 0);
	if ($! != 0) {
		if ($!{EINVAL}) {
		    $AutoLoader::AUTOLOAD = $AUTOLOAD;
		    goto &AutoLoader::AUTOLOAD;
		} else {
		    croak "Your vendor has not defined Device::MX240 macro $constname";
		}
	}
	{
		no strict 'refs';
		# Fixed between 5.005_53 and 5.005_61
		if ($] >= 5.00561) {
		    *$AUTOLOAD = sub () { $val };
		} else {
		    *$AUTOLOAD = sub { $val };
		}
	}
	goto &$AUTOLOAD;
}

Device::MX240->bootstrap($VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Device::MX240 - Perl api to the MX240 (Motorola Instant Messenger)

=head1 SYNOPSIS

	use Device::MX240 qw(open_dev);

	my $done = 0;

	# Ctrl-C/INT handler
	$SIG{INT} = sub { $SIG{INT} = 'IGNORE'; $done++; };

	my $d = open_dev();

	# init device
	$d->write_dev("\xad\xef\x8d");

	while (!$done) {
		my $buf = $d->read_dev();
		if ($buf) {
			# process $buf
			# and use $d->write_dev();
		}
		sleep(1);
		# poll the device
		$d->write_dev("\xad");
	}

	$d->close_dev;

	exit;

=head1 ABSTRACT

This module allows easy communication to and from the MX240.  Also known
as the IMFree.  The device is a wireless instant messenger.
(USB base, 900mhz) libusb and libhid are required for this module to
compile.

=head1 DESCRIPTION

This module has a few functions to interact with the MX240a.  Data is read in
16 byte boundries, and is written in 16 byte boundries.

=head2 CONSTRUCTOR

=over 4

=item C<open_dev>

This function returns a blessed reference to the MX240 device.  The interface
for this function may change in the future to accept a vendor and product id.
You can use the returned object with the following mehods.

=back

=head2 METHODS

=over 4

=item C<read_dev>

	Returns undef when nothing has been read, and 16 bytes of data on success.

=item C<write_dev( $data )>

	Returns number of bytes written, and 0 on failure.

=item C<close_dev>

	Returns true on success and 0 on failure;

=back

=head2 EXPORTS

None by default.  You can export C<open_dev>, C<MX240A_VENDOR>, and C<MX240A_PRODUCT>.

	use Device::MX240 qw(open_dev);

=head1 SEE ALSO

L<http://libusb.sourceforge.net/>, L<http://libhid.alioth.debian.org/>

=head1 AUTHOR

David Davis E<lt>xantus@cpan.orgE<gt>

=head1 CREDITS

Dusty, for the C functions to libhid and libusb.

Please rate this module. L<"http://cpanratings.perl.org/rate/?distribution=Device-MX240">

=head1 COPYRIGHT AND LICENSE

Copyright 2005 by David Davis and Teknikill Software

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
